packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/BasicTypeConverters.kt

```diff
@@ -3,58 +3,107 @@ package expo.modules.kotlin.types
 import com.facebook.react.bridge.Dynamic
 import com.facebook.react.bridge.ReadableArray
 import com.facebook.react.bridge.ReadableMap
+import expo.modules.kotlin.jni.CppType
+import expo.modules.kotlin.jni.JavaScriptObject
+import expo.modules.kotlin.jni.JavaScriptValue
 
-class IntTypeConverter(isOptional: Boolean) : TypeConverter<Int>(isOptional) {
-  override fun convertNonOptional(value: Dynamic): Int = value.asInt()
+class IntTypeConverter(isOptional: Boolean) : DynamicAwareTypeConverters<Int>(isOptional) {
+  override fun convertFromDynamic(value: Dynamic): Int = value.asInt()
+  override fun convertFromAny(value: Any): Int = when (value) {
+    is Double -> value.toInt()
+    else -> value as Int
+  }
+
+  override fun getCppRequiredTypes(): List<CppType> = listOf(CppType.DOUBLE)
 }
 
-class DoubleTypeConverter(isOptional: Boolean) : TypeConverter<Double>(isOptional) {
-  override fun convertNonOptional(value: Dynamic): Double = value.asDouble()
+class DoubleTypeConverter(isOptional: Boolean) : DynamicAwareTypeConverters<Double>(isOptional) {
+  override fun convertFromDynamic(value: Dynamic): Double = value.asDouble()
+  override fun convertFromAny(value: Any): Double = when (value) {
+    is Int -> value.toDouble()
+    else -> value as Double
+  }
+
+  override fun getCppRequiredTypes(): List<CppType> = listOf(CppType.DOUBLE)
 }
 
-class FloatTypeConverter(isOptional: Boolean) : TypeConverter<Float>(isOptional) {
-  override fun convertNonOptional(value: Dynamic): Float = value.asDouble().toFloat()
+class FloatTypeConverter(isOptional: Boolean) : DynamicAwareTypeConverters<Float>(isOptional) {
+  override fun convertFromDynamic(value: Dynamic): Float = value.asDouble().toFloat()
+  override fun convertFromAny(value: Any): Float = when (value) {
+    is Int -> value.toFloat()
+    is Double -> value.toFloat()
+    else -> value as Float
```

Comment:
Wondering (this applies to all numeric converters here: int, float, double) if we could use the Number class directly:
```suggestion
    is Number -> value.toFloat()
    else -> value as Float // this case is actually NaN
```
Because all `Int`, `Float` and `Double` inherit from the `Number` class so we could make it more universal